### PR TITLE
Adds the `PlayerArmorStandManipulate` event

### PR DESCRIPTION
**Pull request by heypr**
- From https://github.com/DenizenScript/Denizen/pull/2600
- Requested by [Darwin](https://discord.com/channels/315163488085475337/1180954233227460749)

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    public PlayerArmorStandManipulateEvent event;
+    public EntityTag entity;
+    public ItemTag armorStandItem;
+    public ItemTag playerItem;
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        if (!runInCheck(path, event.getPlayer().getLocation())) {
+            return false;
+        }
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java:75]: switches for the items & slot would be pretty handy I think

**Review by mcmonkey4eva — APPROVED** [review]

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        if (!path.tryObjectSwitch("armor_stand_item", armorStandItem)) {
+            return false;
+        }
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_19) && !path.tryObjectSwitch("hand", new ElementTag(event.getHand()))) {
+            return false;
+        }
+        if (!path.tryObjectSwitch("player_item", playerItem)) {
+            return false;
+        }
+        if (!path.tryObjectSwitch("slot", new ElementTag(event.getSlot()))) {
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java]: Enum switches should use `runGenericSwitchCheck` (also applies to the other one)

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    //
+    // @Location true
+    //
+    // @Cancellable true
+    //
+    // @Triggers when a player manipulates an armor stand entity.
+    //
+    // @Switch armor_stand_item:<item> to only process the event if the armor stand is holding a specific item.
+    // @Switch hand:<hand> to only process the event if the player is using a specific hand to interact with the armor stand. Available only on MC versions 1.19+.
+    // @Switch player_item:<item> to only process the event if the player is holding a specific item.
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java]: Switch meta should specify it's a matcher, e.g. something like `player_item:<item> to only process the event if the item held by the player matches the specified item matcher.`

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    //
+    // @Triggers when a player manipulates an armor stand entity.
+    //
+    // @Switch armor_stand_item:<item> to only process the event if item held by the armor stand matches the specified item matcher.
+    // @Switch hand:<hand> to only process the event if the player is using a specific hand to interact with the armor stand. Available only on MC versions 1.19+.
+    // @Switch player_item:<item> to only process the event if the item held by the player matches the specified item matcher.
+    // @Switch slot:<slot> to only process the event if the armor stand's item slot that was interacted with is the specified slot.
+    //
+    // @Context
+    // <context.armor_stand_item> returns the ItemTag held by the armor stand.
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java]: Should verify what this item actually is and update meta accordingly, see discussion on [Discord](https://discord.com/channels/315163488085475337/1011496047811506227/1220418650151059538)

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    public PlayerArmorStandManipulateEvent event;
+    public EntityTag entity;
+    public ItemTag armorStandItem;
+    public ItemTag playerItem;
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        if (!runInCheck(path, event.getPlayer().getLocation())) {
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java]: Probably makes more sense to use the armor stand's location here? the `in:` switch with this event would probably be used for stuff like protecting armor stands in a specific area

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        if (!path.tryObjectSwitch("armor_stand_item", armorStandItem)) {
+            return false;
+        }
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_19) && !runGenericSwitchCheck(path,"hand", event.getHand().name())) {
+            return false;
+        }
+        if (!path.tryObjectSwitch("player_item", playerItem)) {
+            return false;
+        }
+        if (!runGenericSwitchCheck(path,"slot", event.getSlot().name())) {
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java]: Missing space between args (on the other `runGenericSwitchCheck` as well)

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    // <context.hand> returns an ElementTag of the hand used by the player to interact with the armor stand, can be either HAND or OFF_HAND. Available only on MC versions 1.19+.
+    // <context.player_item> returns the ItemTag held by the player.
+    // <context.slot> returns an ElementTag of the armor stand's item slot that was interacted with. Valid equipment slot values can be found at <@link url https://hub.spigotmc.org/javadocs/spigot/org/bukkit/inventory/EquipmentSlot.html>.
+    //
+    // @Player Always.
+    // -->
+
+    public PlayerArmorStandManipulateScriptEvent() {
+        registerCouldMatcher("player edits armor stand");
+        registerSwitches("armor_stand_item", "hand", "player_item", "slot");
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java]: Should have some switch/matcher for the armor stand itself as well

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+import com.denizenscript.denizencore.scripts.ScriptEntryData;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
+
+public class PlayerArmorStandManipulateScriptEvent extends BukkitScriptEvent implements Listener {
+
+    // <--[event]
+    // @Events
+    // player edits <'armor_stand'>
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerArmorStandManipulateScriptEvent.java]: this event line is extremely liable to conflict with events, currently it conflicts with 1 (`player edits book`)

**Comment by tal5** [comment]
LGTM, although `manipulates` is a bit inconsistent with other event names imo - now that the event line has a hard-coded `armor stand` (I.e. won't conflict), I'd probably do `edits` or `changes`

**Comment by mcmonkey4eva** [comment]
Maybe `player changes armor stand item from:item to:item` (ie the line is `player changes armor stand item` but swap the player/armorstand _item switches to from/to so it reads really cleanly)

**Review by tal5 — APPROVED** [review]

**Pull request merged**